src/components/DayPickerRangeController.jsx

```diff
@@ -595,39 +595,55 @@ export default class DayPickerRangeController extends React.Component {
   }
 
   addModifier(updatedDays, day, modifier) {
-    const { numberOfMonths, enableOutsideDays } = this.props;
-    const { currentMonth, visibleDays } = this.state;
+    const { numberOfMonths: numberOfVisibleMonths, enableOutsideDays, orientation } = this.props;
+    const { currentMonth: firstVisibleMonth, visibleDays } = this.state;
+
+    let currentMonth = firstVisibleMonth;
+    let numberOfMonths = numberOfVisibleMonths;
+    if (orientation !== VERTICAL_SCROLLABLE) {
+      currentMonth = currentMonth.clone().subtract(1, 'month');
+      numberOfMonths += 2;
+    }
     if (!day || !isDayVisible(day, currentMonth, numberOfMonths, enableOutsideDays)) {
       return updatedDays;
     }
-    let monthIso = toISOMonthString(day);
-    let month = updatedDays[monthIso] || visibleDays[monthIso];
+
     const iso = toISODateString(day);
 
+    let updatedDaysAfterAddition = { ...updatedDays };
     if (enableOutsideDays) {
-      const startOfMonth = day.clone().startOf('month');
-      const endOfMonth = day.clone().endOf('month');
-      if (
-        isBeforeDay(startOfMonth, currentMonth.clone().startOf('month')) ||
-        isAfterDay(endOfMonth, currentMonth.clone().endOf('month'))
-      ) {
-        monthIso = Object.keys(visibleDays).filter(monthKey => (
-          monthKey !== monthIso && Object.keys(visibleDays[monthKey]).indexOf(iso) > -1
-        ))[0];
-        month = updatedDays[monthIso] || visibleDays[monthIso];
-      }
+      const monthsToUpdate = Object.keys(visibleDays).filter(monthKey => (
+        Object.keys(visibleDays[monthKey]).indexOf(iso) > -1
+      ));
+
+      updatedDaysAfterAddition = monthsToUpdate.reduce((days, monthIso) => {
```

Comment:
I think it is complicated to share these methods because a lot of these variables depend on state and/or props. I think the dream scenario is eventually to have a `withModifiers` HOC of some sort that could do a lot of this stuff on its own.